[DebugInfo] Fix handling of @_originallyDefinedIn types

augusto2112 opened the pull request:
Emit an imported declaration for @_originallyDefinedIn under the
    real module that these types live in.

    This patch also changes the mangling for the debugger to respect
    @_originallyDefinedIn, and fixes a bug where @_originallyDefinedIn
    that should be ignored was still being used when mangling.

    rdar://137146961

augusto2112 commented:
https://github.com/swiftlang/llvm-project/pull/9657
@swift-ci test

slavapestov commented on include/swift/AST/ASTMangler.h:
```diff
@@ -348,7 +348,8 @@ class ASTMangler : public Mangler {
 
-  std::string mangleTypeForDebugger(Type decl, GenericSignature sig);
+  std::string mangleTypeForDebugger(Type decl, GenericSignature sig,
+                                    bool RespectOriginallyDefinedIn = false);
```
This is only called in three places. How about instead of adding a parameter here, you just change the three callers to set RespectOriginallyDefinedIn to true or false as appropriate

slavapestov commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -1521,7 +1523,7 @@ createSpecializedStructOrClassType(NominalOrBoundGenericNominalType *Type,
     for (auto Arg : GenericArgs) {
       DebugTypeInfo ParamDebugType;
-      if (Opts.DebugInfoLevel > IRGenDebugInfoLevel::ASTTypes)
+      if (Opts.DebugInfoLevel > IRGenDebugInfoLevel::ASTTypes && !llvm::isa<GenericTypeParamType>(Arg) && !AsForwardDeclarations)
```
`!llvm::isa<GenericTypeParamType>(Arg)` should be written as `!Arg->is<GenericTypeParamType>()`, but that sounds like the wrong condition. What are you trying to determine here?

augusto2112 replied:
Re-examining this the condition was incorrect, I've removed it.

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci test

augusto2112 committed:
[DebugInfo] Fix handling of @_originallyDefinedIn types

Emit an imported declaration for @_originallyDefinedIn under the
real module that these types live in.

This patch also changes the mangling for the debugger to respect
@_originallyDefinedIn, and fixes a bug where @_originallyDefinedIn
that should be ignored was still being used when mangling.

rdar://137146961

augusto2112 commented:
https://github.com/swiftlang/llvm-project/pull/9657
@swift-ci test

augusto2112 commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -2338,6 +2330,106 @@ createSpecializedStructOrClassType(NominalOrBoundGenericNominalType *Type,
+  /// Returns true if the type's mangled name is affected by an
+  /// @_originallyDefinedIn annotation. This annotation can be on the type
+  /// itself, one of its generic arguments, etc.
+  bool containsOriginallyDefinedIn(Type T) {
```
@slavapestov I found the following situations where @_originallyDefinedIn could affect a type's mangled name:

- The type itself being annotated.
- The bound generic arguments being annotated.
-  A typealias local to a function could be affected by the types in the function's generic signature.
- A metatype's type.

Do you know of any other ones?

augusto2112 commented:
https://github.com/swiftlang/llvm-project/pull/9657
@swift-ci test

augusto2112 commented:
https://github.com/swiftlang/llvm-project/pull/9657
@swift-ci test

augusto2112 commented:
https://github.com/swiftlang/llvm-project/pull/9657
@swift-ci test source compatibility

adrian-prantl commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -2338,6 +2330,106 @@ createSpecializedStructOrClassType(NominalOrBoundGenericNominalType *Type,
+    if (auto *MT = llvm::dyn_cast<MetatypeType>(T))
+      if (containsOriginallyDefinedIn(MT->getInstanceType()))
+        return true;
+    ;
```
stray ;

adrian-prantl commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -2338,6 +2330,106 @@ createSpecializedStructOrClassType(NominalOrBoundGenericNominalType *Type,
+
+    // If the type is a bound generic, the type of the substituted generic
+    // arguments might be annotated with @_originallyDefinedIn.
+    if (auto *BGT = llvm::dyn_cast<BoundGenericType>(T))
```
I would have expected this function to be implemented in terms of a type visitor. For example, does this catch a a tuple containing an originallyDefinedIn type?

augusto2112 commented:
@swift-ci test source compatibility

augusto2112 commented:
Closed in favor of https://github.com/swiftlang/swift/pull/78104
